sql/log.h

```diff
@@ -600,9 +600,12 @@ class binlog_cache_mngr;
 class binlog_cache_data;
 struct rpl_gtid;
 struct wait_for_commit;
+class Binlog_commit_by_rotate;
 
 class MYSQL_BIN_LOG: public TC_LOG, private Event_log
 {
+  friend Binlog_commit_by_rotate;
```

Comment:
As I see there is only one instance of the class declared `static` in `log.cc` and it is always accessed
within `MYSQL_BIN_LOG::method`:s. Except one time in ` binlog_cache_reserved_size()`.
Of course locally `Binlog_commit_by_rotate` is a pure sub-class in the sense that its instance (one) can not (does not have to) exist without  `mysql_bin_log` instance.
I wonder why the class was not made as internal to  `MYSQL_BIN_LOG`. The latter function then would access `mysql_bin_log` instance instead of the current way   of  `return binlog_commit_by_rotate.get_reserved_size()`.

Reply:
I didn't make it as a member of MYSQL_BIN_LOG, because MYSQL_BIN_LOG is shared between binlog and relay log. But binlog_commit_by_rotate is not really necessary for relay log.